File: user-scripts/javascript/templates/basic_script.js
#!/usr/bin/env node

/**
 * Task: [Brief description of what this script does]
 * Created: [YYYY-MM-DD]
 * Author: [Creator name/identifier]
 * 
 * Description:
 *   [Detailed description of the script's purpose, inputs, outputs, and behavior]
 * 
 * Usage:
 *   node basic_script.js [arguments]
 *   
 *   Arguments:
 *     --input, -i:   Input file or data source
 *     --output, -o:  Output file (default: stdout)
 *     --config, -c:  Configuration file path
 *     --verbose, -v: Enable verbose logging
 *     --dry-run:     Show what would be done without executing
 *     --help, -h:    Show this help message
 * 
 *   Examples:
 *     node basic_script.js --input data.json --output results.json
 *     node basic_script.js --input data.csv --verbose
 *     node basic_script.js --help
 * 
 * Dependencies:
 *   - Node.js 16+
 *   - Required packages: [list packages]
 * 
 * Notes:
 *   - [Any important notes or limitations]
 *   - [Performance considerations]
 *   - [Known issues or workarounds]
 */

const fs = require('fs').promises;
const path = require('path');
const { program } = require('commander');

// Configuration
const PROJECT_ROOT = path.resolve(__dirname, '../../../..');
const LOG_DIR = path.join(PROJECT_ROOT, 'user-scripts/shared/logs');

/**
 * Custom error class for script-specific errors
 */
class ScriptError extends Error {
    constructor(message) {
        super(message);
        this.name = 'ScriptError';
    }
}

/**
 * Simple logging utility
 */
class Logger {
    constructor(verbose = false) {
        this.verbose = verbose;
        this.logFile = path.join(LOG_DIR, 'script.log');
    }

    async _writeLog(level, message) {
        const timestamp = new Date().toISOString();
        const logEntry = `${timestamp} - ${level} - ${message}\n`;
        
        // Write to console
        if (level === 'ERROR' || this.verbose || level === 'INFO') {
            console.log(`${timestamp} - ${level} - ${message}`);
        }
        
        // Write to file
        try {
            await fs.appendFile(this.logFile, logEntry);
        } catch (err) {
            console.error('Failed to write to log file:', err.message);
        }
    }

    async info(message) {
        await this._writeLog('INFO', message);
    }

    async debug(message) {
        if (this.verbose) {
            await this._writeLog('DEBUG', message);
        }
    }

    async error(message) {
        await this._writeLog('ERROR', message);
    }

    async warn(message) {
        await this._writeLog('WARN', message);
    }
}

/**
 * Load configuration from file or use defaults
 */
async function loadConfig(configPath = null) {
    const defaultConfig = {
        timeout: 30000,
        maxRetries: 3,
        batchSize: 100,
    };

    if (configPath) {
        try {
            const configData = await fs.readFile(configPath, 'utf8');
            const userConfig = JSON.parse(configData);
            Object.assign(defaultConfig, userConfig);
            logger.info(`Loaded configuration from ${configPath}`);
        } catch (err) {
            logger.warn(`Failed to load config from ${configPath}: ${err.message}`);
        }
    }

    return defaultConfig;
}

/**
 * Validate input arguments and files
 */
async function validateInputs(options) {
    try {
        await fs.access(options.input);
    } catch (err) {
        throw new ScriptError(`Input file does not exist: ${options.input}`);
    }

    if (options.output) {
        const outputDir = path.dirname(options.output);
        try {
            await fs.mkdir(outputDir, { recursive: true });
            logger.debug(`Ensured output directory exists: ${outputDir}`);
        } catch (err) {
            throw new ScriptError(`Failed to create output directory: ${err.message}`);
        }
    }
}

/**
 * Main data processing function
 */
async function processData(inputPath, config) {
    logger.info(`Processing data from ${inputPath}`);
    
    try {
        const results = [];
        
        // Read input file
        const inputData = await fs.readFile(inputPath, 'utf8');
        
        // Example processing: split by lines and process each
        const lines = inputData.split('\n').filter(line => line.trim());
        
        for (let i = 0; i < lines.length; i++) {
            try {
                const processedItem = {
                    lineNumber: i + 1,
                    content: lines[i].trim(),
                    processedAt: new Date().toISOString(),
                    // TODO: Add your processing logic here
                };
                
                results.push(processedItem);
                
                if ((i + 1) % config.batchSize === 0) {
                    logger.info(`Processed ${i + 1} items`);
                }
                
            } catch (err) {
                logger.error(`Error processing line ${i + 1}: ${err.message}`);
                continue;
            }
        }
        
        logger.info(`Successfully processed ${results.length} items`);
        return results;
        
    } catch (err) {
        throw new ScriptError(`Failed to process data: ${err.message}`);
    }
}

/**
 * Save results to file or stdout
 */
async function saveResults(results, outputPath = null) {
    const outputData = {
        timestamp: new Date().toISOString(),
        totalItems: results.length,
        data: results
    };
    
    const jsonOutput = JSON.stringify(outputData, null, 2);
    
    if (outputPath) {
        await fs.writeFile(outputPath, jsonOutput, 'utf8');
        logger.info(`Results saved to ${outputPath}`);
    } else {
        console.log(jsonOutput);
    }
}

/**
 * Main script execution function
 */
async function main() {
    let logger;
    
    try {
        // Setup command line interface
        program
            .name('basic_script.js')
            .description('[Script description]')
            .requiredOption('-i, --input <path>', 'Input file or data source')
            .option('-o, --output <path>', 'Output file (default: stdout)')
            .option('-c, --config <path>', 'Configuration file path')
            .option('-v, --verbose', 'Enable verbose logging')
            .option('--dry-run', 'Show what would be done without executing')
            .helpOption('-h, --help', 'Show this help message');

        program.parse();
        const options = program.opts();
        
        // Initialize logger
        logger = new Logger(options.verbose);
        
        // Ensure log directory exists
        await fs.mkdir(LOG_DIR, { recursive: true });
        
        logger.debug('Script started with options:', JSON.stringify(options, null, 2));
        
        // Load configuration
        const config = await loadConfig(options.config);
        logger.debug(`Using configuration: ${JSON.stringify(config, null, 2)}`);
        
        // Validate inputs
        await validateInputs(options);
        
        // Dry run check
        if (options.dryRun) {
            logger.info('DRY RUN: Would process data without making changes');
            logger.info(`Input: ${options.input}`);
            logger.info(`Output: ${options.output || 'stdout'}`);
            return 0;
        }
        
        // Process data
        const results = await processData(options.input, config);
        
        // Save results
        await saveResults(results, options.output);
        
        logger.info('Script completed successfully');
        return 0;
        
    } catch (err) {
        if (logger) {
            if (err instanceof ScriptError) {
                logger.error(`Script error: ${err.message}`);
            } else {
                logger.error(`Unexpected error: ${err.message}`);
                if (err.stack) {
                    logger.debug(`Stack trace: ${err.stack}`);
                }
            }
        } else {
            console.error('Error:', err.message);
        }
        return 1;
    }
}

// Handle process signals
process.on('SIGINT', async () => {
    if (logger) {
        logger.info('Script interrupted by user');
    }
    process.exit(1);
});

process.on('uncaughtException', async (err) => {
    if (logger) {
        logger.error(`Uncaught exception: ${err.message}`);
    }
    console.error('Uncaught Exception:', err);
    process.exit(1);
});

// Execute main function if script is run directly
if (require.main === module) {
    main().then(exitCode => {
        process.exit(exitCode);
    }).catch(err => {
        console.error('Fatal error:', err);
        process.exit(1);
    });
}

module.exports = {
    main,
    processData,
    saveResults,
    loadConfig,
    validateInputs,
    ScriptError,
    Logger
};

File: user-scripts/python/templates/basic_script.py
#!/usr/bin/env python3
"""
Task: [Brief description of what this script does]
Created: [YYYY-MM-DD]
Author: [Creator name/identifier]

Description:
    [Detailed description of the script's purpose, inputs, outputs, and behavior]

Usage:
    python basic_script.py [arguments]
    
    Arguments:
        arg1: Description of first argument
        arg2: Description of second argument
    
    Examples:
        python basic_script.py --input data.csv --output results.json
        python basic_script.py --help

Dependencies:
    - Python 3.8+
    - Required packages: [list packages]
    
Notes:
    - [Any important notes or limitations]
    - [Performance considerations]
    - [Known issues or workarounds]
"""

import argparse
import logging
import os
import sys
from pathlib import Path
from typing import Optional, List, Dict, Any

# Add the project root to Python path for importing MCP modules if needed
PROJECT_ROOT = Path(__file__).parent.parent.parent.parent
sys.path.insert(0, str(PROJECT_ROOT / "src"))

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
    handlers=[
        logging.FileHandler(PROJECT_ROOT / "user-scripts/shared/logs/script.log"),
        logging.StreamHandler(sys.stdout)
    ]
)
logger = logging.getLogger(__name__)


class ScriptError(Exception):
    """Custom exception for script-specific errors."""
    pass


def parse_arguments() -> argparse.Namespace:
    """Parse command line arguments."""
    parser = argparse.ArgumentParser(
        description="[Script description]",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
    python basic_script.py --input data.csv
    python basic_script.py --input data.csv --output results.json --verbose
        """
    )
    
    parser.add_argument(
        '--input', '-i',
        type=str,
        required=True,
        help='Input file or data source'
    )
    
    parser.add_argument(
        '--output', '-o',
        type=str,
        help='Output file (default: stdout)'
    )
    
    parser.add_argument(
        '--config', '-c',
        type=str,
        help='Configuration file path'
    )
    
    parser.add_argument(
        '--verbose', '-v',
        action='store_true',
        help='Enable verbose logging'
    )
    
    parser.add_argument(
        '--dry-run',
        action='store_true',
        help='Show what would be done without executing'
    )
    
    return parser.parse_args()


def load_config(config_path: Optional[str] = None) -> Dict[str, Any]:
    """Load configuration from file or use defaults."""
    default_config = {
        'timeout': 30,
        'max_retries': 3,
        'batch_size': 100,
    }
    
    if config_path and os.path.exists(config_path):
        try:
            import json
            with open(config_path, 'r') as f:
                user_config = json.load(f)
            default_config.update(user_config)
            logger.info(f"Loaded configuration from {config_path}")
        except Exception as e:
            logger.warning(f"Failed to load config from {config_path}: {e}")
    
    return default_config


def validate_inputs(args: argparse.Namespace) -> None:
    """Validate input arguments and files."""
    if not os.path.exists(args.input):
        raise ScriptError(f"Input file does not exist: {args.input}")
    
    if args.output:
        output_dir = os.path.dirname(args.output)
        if output_dir and not os.path.exists(output_dir):
            os.makedirs(output_dir, exist_ok=True)
            logger.info(f"Created output directory: {output_dir}")


def process_data(input_path: str, config: Dict[str, Any]) -> List[Dict[str, Any]]:
    """
    Main data processing function.
    
    Args:
        input_path: Path to input file
        config: Configuration dictionary
        
    Returns:
        List of processed data items
        
    Raises:
        ScriptError: If processing fails
    """
    logger.info(f"Processing data from {input_path}")
    
    try:
        # TODO: Implement your data processing logic here
        results = []
        
        # Example processing loop
        with open(input_path, 'r') as f:
            for line_num, line in enumerate(f, 1):
                try:
                    # Process each line/item
                    processed_item = {
                        'line_number': line_num,
                        'content': line.strip(),
                        'processed_at': str(logger.handlers[0].formatter.formatTime(
                            logging.LogRecord('', 0, '', 0, '', (), None)
                        ))
                    }
                    results.append(processed_item)
                    
                    if line_num % config['batch_size'] == 0:
                        logger.info(f"Processed {line_num} items")
                        
                except Exception as e:
                    logger.error(f"Error processing line {line_num}: {e}")
                    continue
        
        logger.info(f"Successfully processed {len(results)} items")
        return results
        
    except Exception as e:
        raise ScriptError(f"Failed to process data: {e}")


def save_results(results: List[Dict[str, Any]], output_path: Optional[str] = None) -> None:
    """Save results to file or stdout."""
    import json
    
    output_data = {
        'timestamp': str(logger.handlers[0].formatter.formatTime(
            logging.LogRecord('', 0, '', 0, '', (), None)
        )),
        'total_items': len(results),
        'data': results
    }
    
    if output_path:
        with open(output_path, 'w') as f:
            json.dump(output_data, f, indent=2)
        logger.info(f"Results saved to {output_path}")
    else:
        print(json.dumps(output_data, indent=2))


def main() -> int:
    """Main script execution function."""
    try:
        # Parse arguments
        args = parse_arguments()
        
        # Set log level
        if args.verbose:
            logging.getLogger().setLevel(logging.DEBUG)
            logger.debug("Verbose logging enabled")
        
        # Load configuration
        config = load_config(args.config)
        logger.debug(f"Using configuration: {config}")
        
        # Validate inputs
        validate_inputs(args)
        
        # Dry run check
        if args.dry_run:
            logger.info("DRY RUN: Would process data without making changes")
            logger.info(f"Input: {args.input}")
            logger.info(f"Output: {args.output or 'stdout'}")
            return 0
        
        # Process data
        results = process_data(args.input, config)
        
        # Save results
        save_results(results, args.output)
        
        logger.info("Script completed successfully")
        return 0
        
    except ScriptError as e:
        logger.error(f"Script error: {e}")
        return 1
    except KeyboardInterrupt:
        logger.info("Script interrupted by user")
        return 1
    except Exception as e:
        logger.exception(f"Unexpected error: {e}")
        return 1


if __name__ == "__main__":
    exit_code = main()
    sys.exit(exit_code)